### PR TITLE
Gaussian processes

### PR DESCRIPTION
**Pull request by bwengals**
Adding a basic set of covariance functions / kernels for doing Gaussian Process regression.  They follow a similar API as GPy and GPFlow.  Some main differences are:
  - Covariance function objects allow for multiplication and addition with a scalar (or PyMC3 variable) or a numpy array of the right size.
  - Covariance functions aren't named, their parameters are PyMC3 variables which have names.

At this point, Gibbs and Linear aren't correct yet.  Hope to eventually have a fairly complete suite of covariance functions.

This WIP PR also includes some notebook examples showing how to use them, and an example of fitting a GP.  

Please, feedback and suggestions welcome!

### CODE REVIEW

File: pymc3/__init__.py [hunk 1]
```diff
@@ -23,6 +23,7 @@
 from .tests import test
 
 from . import glm
+from . import gp
```
Comment (twiecki) [inline pymc3/__init__.py]: Can we move this to `pymc3.models`?

**Comment by twiecki** [comment]
I've only skimmed so far but this is very exciting. Really like the compositionality of kernels. Some high-level points:
- [x] Should move to `pymc3.models.gp`.
- [ ] The examples should live in `docs/sources/notebooks` and we should add a separate category in `docs/sources/examples.rst` for them.
- [x] Should we renamed `covariance` to `kernel`? I feel like the latter is more commonly used in this context and has less risk of being confused with covariance used elsewhere. Then again, it directly links to the covariance being passed to `MvNormal` so maybe it's actually more consistent that way.
- [ ] How does NUTS perform?
- [ ] Needs unit-tests.
- [x] Rebase to `3.1` branch, soon to be merged into master.

File: pymc3/distributions/multivariate.py [hunk 2]
```diff
@@ -8,6 +8,7 @@
 import theano.tensor as tt
 
 from scipy import stats
+
```
Comment (twiecki) [inline pymc3/distributions/multivariate.py]: Changed permissions.

File: pymc3/gp/cov.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+            Zs = tt.sum(tt.square(Z), 1)
+            return -2.0 * tt.dot(X, tt.transpose(Z)) +\
+                   tt.reshape(Xs, (-1, 1)) + tt.reshape(Zs, (1, -1))
+
+    def euclidean_dist(self, X, Z):
+        r2 = self.square_dist(X, Z)
+        return tt.sqrt(r2 + 1e-12)
+
+
+class ExpQuad(Stationary):
```
Comment (twiecki) [inline pymc3/gp/cov.py:122]: Doc-strings, can we copy & paste those from `GPy`?

File: pymc3/__init__.py (hunk 1 above)
Comment (bwengals) [reply pymc3/__init__.py]: Absolutely.  I wasn't sure where things should go

File: pymc3/gp/cov.py (hunk 3 above)
Comment (bwengals) [reply pymc3/gp/cov.py:122]: These are based on [GPFlow](https://github.com/GPflow/GPflow/blob/master/GPflow/kernels.py).  It was easy to adapt their tensorflow to theano.  For most of them it should be a copy/paste job.

**Comment by bwengals** [comment]
- Sounds good
- Sounds good
- This I was on the fence with too, for the exact reasons you mentioned.  They are usually called kernels, but they literally are covariances.  I'm okay with either naming convention though.
- Not really sure yet.  I found NUTS would hang or get stuck sometimes.  Slice seems to work the best.  This is just a first impression though.
- Agree
- Sounds good

Awesome!

**Comment by twiecki** [comment]
OK, NUTS might need some help tuning, we can leave that for now, it's a more general problem we're working on.

File: pymc3/__init__.py (hunk 1 above)
Comment (fonnesbeck) [reply pymc3/__init__.py]: I think `gp` is fine. I don't think we want both `pymc3.model` and `pymc3.models`.

**Comment by fonnesbeck** [comment]
OK, I see on 3.1 we now have a `models` submodule. I will provide feedback there; I don't think that's a good idea.

**Comment by twiecki** [comment]
@fonnesbeck The `models` submodule or placing GP there? I actually have my concerns on the former as well. 

Maybe we can make progress this way too: Thinking about this PR very narrowly, it does not implement GPs, but rather functionality to construct covariance matrices with certain structure. These just happen to be used in GPs. @bwengals thoughts?

**Comment by fonnesbeck** [comment]
I think GPs are unique enough (in the sense that they are models of functions) to have their own submodule. I'd be open to `pymc3.np`, or similar, thinking ahead to when we might have Dirichlet processes, etc.

**Comment by twiecki** [comment]
Definitely, just this isn't GP, it's covariance matrix generators. Just thinking of how to name / structure this. I'm also not sure Dirichlet Process would be a submodule, it would just be a another distribution. And we already have an example of a DPGMM. Since we're not providing models but building blocks for models, the naming is a bit different.

**Comment by fonnesbeck** [comment]
`kernels` might work. I would favor something along the lines of `pymc3.gp` alongside `pymc3.kernels`.

**Comment by bwengals** [comment]
I'm thinking that including a `gp` submodule where the kernels/covariances would be appropriate if there was some convenience function modeled after `glm`.  Perhaps something like:
```
with pm.Model() as model2:
    l = pm.HalfNormal('l')
    cov_f = pm.gp.Matern52(1,l)
    gp1 = pm.gp.GP('gp1', mean=0, latent_cov=cov_f, likelihood='poisson', data=X, observed=y)
```
for a Poisson model, or 
```
with pm.Model() as model2:
    ...
    gp2 = pm.gp.GP('gp2', mean=mean_func, latent_cov=cov_f, noise_cov=cov_n, likelihood='normal', observed=y, data=X, predict=Z)
```
for a normal model.  I started this functionality, but got a bit bogged down in the details and decided to hold off for the time being.

If not a `gp` directory, maybe use a `kernels` or `cov_funcs` directory in the meantime?  It's also possible that as more kernels are added, cov.py will need to be broken into multiple files.

**Comment by twiecki** [comment]
I merged 3.1 into master, so you can just rebase off of master again.

I think such functionality would be great. I'm not opposed to a gp subdirectory, just exploring ideas. If that gets added it sounds right to have a gp submodule. We can leave that for now.

**Comment by twiecki** [comment]
Also, the likelihood should probably be a dist:
```python
with pm.Model() as model2:
    l = pm.HalfNormal('l')
    cov_f = pm.gp.Matern52(1,l)
    gp1 = pm.gp.GP('gp1', mean=0, latent_cov=cov_f, likelihood=pm.Poisson, data=X, observed=y)
```

**Comment by twiecki** [comment]
@bwengals I think the only things left then are unittests and doc-stirngs.

**Comment by bwengals** [comment]
Hi @twiecki, haven't had time to work on this lately, sorry about that!  I should be able to finish these two things up within the next couple days.   

I have a question though.  How do you guys want to handle additions to the GP library after the covariance functions get merged into master?  Should new PRs be opened for each additional feature?  Or should development continue on a GP branch?

**Comment by twiecki** [comment]
@bwengals Sounds good. Regarding PRs: it's really up to you. If you're not super sure about the design and want to experiment perhaps a separate branch is the way to go.

**Comment by fonnesbeck** [comment]
I would merge new functionality into master as they are ready for use, and keep work-in-progress on a separate branch. We might as well get new features in the hands of users on the master branch if they are working so that they can be more thoroughly tested. Thanks for your work on this.

**Comment by bwengals** [comment]
Looking forward to getting this in.  Thanks guys, that sounds good.

**Comment by twiecki** [comment]
@bwengals Let us know when you want us to take another look.

**Comment by bwengals** [comment]
will do

**Comment by bwengals** [comment]
@twiecki @fonnesbeck I think things are at a good stopping point for the time being.  Documentation looks good to me at least.  Unit tests are outstanding, but the covariance notebook sort of serves this purpose.  Please, let me know if you'd like any changes, particularly to the examples.  I'm not sure if everything I'm doing in there is the "best" PyMC3 way to do it. 

Oh, what was the decision on structure?  ```pm.gp``` next to ```pm.kernels```? Or ```pm.gp.cov```?  Also, kernel vs. covariance function?

File: docs/source/examples.rst [hunk 4]
```diff
@@ -37,6 +36,14 @@ GLM
    notebooks/GLM-poisson-regression.ipynb
    notebooks/hierarchical_partial_pooling.ipynb
 
+GP
+==
+
+.. toctree::
+   notebooks/GP-introduction.ipynb
+   notebooks/GP-covariances.ipynb
+   notebooks/GP-smoothing.ipynb
```
Comment (twiecki) [inline docs/source/examples.rst:45]: I don't think this NB is added.

**Comment by twiecki** [comment]
@bwengals Looks good. lets stick with `pm.gp` and `covariance`, as you have it now.
Note that the permissions of `multivariate.py` need to be still be fixed (reverted) and a NB seems to miss.

**Comment by twiecki** [comment]
Also need to fix headings in GP-introduction.ipynb. First, it needs a single header at the top (with `#`), then all following headings should be sub-headings (i.e. `##` or more).

**Comment by bwengals** [comment]
Sorry about that!  I *think* the issues you mentioned are resolved now, after a merging ```master``` into ```gaussian-processes```.

**Comment by twiecki** [comment]
```
+++[[ true == \t\r\u\e ]]
+++. ./scripts/lint.sh
++++pylint --rcfile=.pylintrc pymc3
************* Module pymc3.gp.cov
pymc3/gp/cov.py:2: [W0611(unused-import), ] Unused import copy
pymc3/gp/cov.py:4: [W0611(unused-import), ] Unused import theano****
```

**Comment by twiecki** [comment]
Also, now two files changed permissions from before. they should be 755.

File: pymc3/distributions/multivariate.py (hunk 2 above)
Comment (twiecki) [inline pymc3/distributions/multivariate.py]: e.g. this file: pymc3/distributions/multivariate.py 100755 → 100644

**Comment by bwengals** [comment]
Which files?  How/where are you seeing the file permission issue?  Thanks for your patience while I'm being a noob.

**Comment by twiecki** [comment]
On github, if you switch to the file view of your PR. multivariate.py

**Comment by bwengals** [comment]
Ok, fixed the issue with ```multivariate.py```.  Thanks for being patient.  Over the last few commits, I added unit tests, had some issues with getting them to pass in the python3.6 build, then got that fixed.  As far as I can tell things look good.

**Comment by twiecki** [comment]
@bwengals Many thanks, this is a great addition!

**Comment by bwengals** [comment]
@twiecki No thank you!  I look forward to contributing more.

**Pull request merged**